numba/cpython/hashing.py

```diff
@@ -47,8 +47,23 @@
 
 @overload(hash)
 def hash_overload(obj):
+    objty = obj
+    err_msg = f"unhashable type: '{objty}'"
+    attempt_generic_msg = ("No __hash__ is defined for object of type "
+                           f"'{objty}' and a generic hash() cannot be "
+                           "performed as there is no suitable object "
+                           "represention in Numba compiled code!")
+
     def impl(obj):
-        return obj.__hash__()
+        if hasattr(obj, '__hash__'):
+            hash_func = getattr(obj, '__hash__')
+            hash_value = hash_func()
+            if hash_value is not None:
```

Comment:
`__hash__` should never return non-integer according to python. Why not let it raise?

Reply:
Indeed, thanks for picking this up, I think it may be just working around a bug in `getattr`. Calls like `getattr(obj, method)` return in a `BoundFunction` against the `obj`, this probably needs to go through another round of resolution to provide the actual type/function. I'll take a look.

Reply:
I think this was caused by the implementations using `@overload_method(<type>, '__hash__')` where they should have used `@overload_attribute(<type>, '__hash__')`.